### ISSUE TITLE
Resource limits cause bogus resource requests.

### ISSUE THREAD

**Issue by andreas-kupries** [cloudfoundry-incubator/quarks-operator#1114]
**Describe the bug**

Setting a memory limit on a job not only generates a kube-level memory limit (:heavy_check_mark:), it __also generates__ an identical-sized memory request (:boom:). In other words, when I ask to to tell kube to keep the job's memory less than X the operator also forces the memory to be at least X. That is bogus.

**To Reproduce**

Deployed kubecf `0.2.0-1739.gfa5c9166-dirty` on minikube 1.8.2.
This used `cf-operator-6.0.2+0.g4eb7fc9f`

The deployment was modified using the ops patch
```
    - type: replace
      path: /instance_groups/name=nats/jobs/name=nats-tls/properties?/quarks/bpm/processes/name=nats-tls/limits/memory
      value: 40Mi
```
which limits the nats:nat-tls job to 40Mi.

In the rendered helm chart (secret `ops-with`) I find

```
- azs: null
  env: (...)
  instances: 1
  jobs:
  (...)
  - name: nats-tls
    properties:
      (...)
      quarks:
        bpm:
	  (...)
          processes:
          - hooks: {}
            limits:
              memory: 40Mi
            name: nats-tls
            unsafe: {}
        (...)
  (... other jobs)
  name: nats
  (...)
```

That looks ok, just a memory limit, no request in sight.

However, when looking at the nats pod itself I find
```
apiVersion: v1
kind: Pod
metadata:
  name: nats-0
  namespace: kubecf
spec:
  affinity:
  containers:
  (...)
  - (...)
    name: nats-tls-nats-tls
    resources:
      limits:
        memory: 40Mi
      requests:
        memory: 40Mi
    (...)
  (...)
```

My limit is present, good. The memory request however, not good, was not asked for.

**Expected behavior**

I expect to not see a memory request when I asked just for a limit.

**Environment**

  - cf-operator 6.0.2+0.g4eb7fc9f
  - kubecf 0.2.0-1739.gfa5c9166-dirty

**Comment by manno** [comment]
@andreas-kupries I couldn't find any code that would duplicate this. However looking into the Kubernetes docs I found https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/#what-if-you-specify-a-container-s-limit-but-not-its-request

According to https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-resource-requests-and-limits

> It is important to remember that the limit can never be lower than the request. If you try this, Kubernetes will throw an error and won’t let you run the container.

So I think this is the expected behavior.

**Comment by viovanov** [comment]
@andreas-kupries this is not a type of behavior that the operator should override.
We need to provide default values for requests as well in kubecf. These should be much much lower (maybe a 10th of the limit size).

**Comment by andreas-kupries** [comment]
@manno Thank you for looking into this. I have to admit, I did not expect k8s itself to be this ... nonsensical.

To recap the info from the referenced kube docs, explicit and infered:

  - The default request (min) is 256Mi.
  - The default limit (max) is 512Mi.
  - When setting just the request (min) to some X the max is left untouched and at the default.
  - When setting just the limit (max) to some X the min is made identical.

It is that last which feels nonsensical to me. In the interest of symmetry I would have thought to leave the min untouched.

Well, except if the new max is less than the default min, then I can understand forcing a new min, ensuring the `min <= max` condition.

But doing that when I set a max larger than the default min ?

Hm. My example was indeed for a case where the new max was less than the default min, so that is square where I said to understand. Now that I know that there is a default min which is not zero (as I had thought). So the case of the problematic behaviour for new max larger than default min is actually not proven. Might have to see what @jandubois experiences as he tests cloudfoundry-incubator/kubecf/pull/1258.

Looking over the kubecf limits it seems that most of them are under the default min actually, with a few exceptions for the heavy-weight containers (garden, rep, and a few others).

Ok. Lets go with setting our own requests (min) as well. That seems to be the only way forward and around the issue.
